clang/lib/CodeGen/CGCall.cpp

```diff
@@ -1836,6 +1837,50 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
          Module.getLangOpts().Sanitize.has(SanitizerKind::Return);
 }
 
+llvm::fp::FPAccuracy convertFPAccuracy(StringRef FPAccuracy) {
+  StringRef AccuracyVal;
+  if (FPAccuracy == "high")
+    return llvm::fp::FPAccuracy::High;
+  else if (FPAccuracy == "medium")
+    return llvm::fp::FPAccuracy::Medium;
+  else if (FPAccuracy == "low")
+    return llvm::fp::FPAccuracy::Low;
+  else if (FPAccuracy == "sycl")
+    return llvm::fp::FPAccuracy::SYCL;
+  else if (FPAccuracy == "cuda")
+    return llvm::fp::FPAccuracy::CUDA;
+  else
+    llvm_unreachable("Unexpected type for FPAccuracy");
+}
+
+void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
+    StringRef Name, llvm::AttrBuilder &FuncAttrs, unsigned ID,
+    const llvm::Type *FuncType) {
+  // First check that a specific accuracy hasn't been requested for this
+  // function.
+  if (!getLangOpts().FPAccuracyFuncMap.empty()) {
+    auto FuncMapIt = getLangOpts().FPAccuracyFuncMap.find(Name.str());
+    if (FuncMapIt != getLangOpts().FPAccuracyFuncMap.end()) {
+      StringRef FPAccuracyVal = llvm::fp::getAccuracyForFPBuiltin(
+          ID, FuncType, convertFPAccuracy(FuncMapIt->second));
+      assert(!FPAccuracyVal.empty() && "A valid accuracy value is expected");
+      FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    }
+  }
+  if (!getLangOpts().FPAccuracyVal.empty()) {
```

Comment:
Should this be an else if?

Reply:
No. We need to check both. The first if is to treat the setting of the fp-accuracy with no specific functions, the second if is to treat the values of accuracy for specific functions. And in the command line we can have both: -ffp-accuracy=va1l -ffp-accuracy=val2:[f1,f2].

Reply:
I'm a little confused. IIUC, if `FPAccuracyFuncMap` is not empty in first `if`, you are adding to `FuncAttrs`. In this `if ` however, you are checking that the size of  `FuncAttrs` is 0 before updating it, which means you won't add the attribute again right? Am I misunderstanding something?